Fix issue 9029: make all types (including basic) match to alias template parameters with priority MATCH.convert.

FeepingCreature opened the pull request:
This may break code. The point of this PR is partially to determine if it does, and if so what.

This is not done by inserting basic types into the symbol table, but simply passing them through in the compiler as objects. This is the main reason I am unconfident about this PR.

FeepingCreature committed:
Fix issue 9029: make types match to alias template parameters with priority MATCH.convert.

FeepingCreature commented:
Hm, this seems to need some fixes. Should it go behind `-experimental or should I do a full DIP writeup?

FeepingCreature commented:
See https://github.com/dlang/phobos/compare/master...FeepingCreature:fix/issue-9029-fixes for required Phobos changes.

jacob-carlborg commented on test/compilable/ddocunittest.d:
```diff
@@ -345,6 +345,8 @@ unittest { foo9757(); bar9757(); }
 ///
 unittest { bar9757(); foo9757(); }
 
+import std.functional : binaryFun;
```
Phobos should ideally not be used in the tests.

FeepingCreature replied:
Oh oops. Hm.

FeepingCreature replied:
Fixed?

jacob-carlborg commented:
No other tests?

FeepingCreature commented:
I should add one, shouldn't I. Hang on.

FeepingCreature commented:
I put it behind `-preview=allowTypesForTemplateAlias`.

jacob-carlborg commented:
Why is the flag necessary?

FeepingCreature commented:
Because it potentially "breaks" existing code due to templates matching that didn't used to match, making template instantiation ambiguous. You can see instances of that in the Phobos PR.

jacob-carlborg commented:
It might be worth having an explicit changelog entry for this enhancement.

TurkeyMan commented:
Ethan will be so happy!

Geod24 commented on src/dmd/dtemplate.d:
```diff
@@ -5697,6 +5697,10 @@ extern (C++) final class TemplateAliasParameter : TemplateParameter
                  *  template X(T) {}        // T => sa
                  */
             }
+            else if (global.params.allowTypesForTemplateAlias && sa && cast(Type) sa)
```
`sa` is an `extern(C++)` class, hence the third condition (`cast(Type) sa`) does not do a dynamic cast, but a reinterpret one.

FeepingCreature replied:
Oh f*ck.

FeepingCreature replied:
Better?

FeepingCreature commented:
Added a changelog entry, fixed a rather embarrassing bug meaning that the alias check matched on every possible parameter, which reminds me that I can't think of any type, value or other thing that alias *doesn't* match now.

wilzbach commented:
Now that the Phobos PR is merged, how about temporarily enabling this by default and checking the output of the project tester?
Because maybe we don't need a preview flag at all.

FeepingCreature commented:
Optimistic! Let's try it!

FeepingCreature commented:
_Damn_. That's _so_ close.

JinShil commented:
> Can anyone poke the Weka guys about looking at this?

Create an issue in their repository.

JohanEngelen commented on changelog/template_alias_matches_basic_types.dd:
```diff
@@ -0,0 +1,22 @@
+variadic parameters already matched basic types. This led to strange replacements for `alias` such as
+`template Template(T...) if (T.length == 1)`. Those replacements are no longer necessary.
+
+As a consequence, in some cases templates will match that did not used to. As this may break existing code,
```
Extra points for showing what cases break, and _how to fix it_. :)

FeepingCreature replied:
Better?

JohanEngelen replied:
thanks. Perhaps add explicitly that the code would break for delegates.
I guess things can break in many different ways. Regardless, the example shows the type of fix that is required.

FeepingCreature commented:
Now that I've fixed issues in Phobos as well as Mecca relating to this, some thoughts.

There was a happy coincidence with alias in that "the sort of things that could be passed to alias" were exactly "the sort of things you might want to, _in some abstract sense_, call." in other words, lambdas, nested functions, structs, and strings, but specifically _not_ things like delegate types. This allowed for easy overloading between `alias A` and `T`, where `T` would represent a runtime callable parameter and `A` a compiletime callable parameter. However, this was of course coincidental.

Now that `alias` can match every type, this is awkward. There is no check in Phobos `std.traits` that encapsulates what `alias` used to do. One might reasonably expect `isCallable` to do the job, but for unclear reasons it is `true` for `void delegate()`, which is _not_ callable in the "can put () on it" sense. And of course, structs complicate this further by being either compiletime or runtime callable.

Is `__traits(compiles, A(...))` sufficient? I'm not sure.

jacob-carlborg commented:
> edit: facepalm it's using the tagged version of mecca...

Yeah, they need to create a new tag.

jacob-carlborg commented:
When this is merge, please make sure the commits are squashed.

jacob-carlborg commented:
> One might reasonably expect isCallable to do the job, but for unclear reasons it is true for void delegate(), which is not callable in the "can put () on it" sense

Why is `void delegate()` not callable?

wilzbach commented:
@FeepingCreature you are already a member of the Buildkite Organization. This means you can login there and then just click on rebuild for the individual test that failed. Saves you a few commits:

![image](https://user-images.githubusercontent.com/4370550/57613531-63f3f080-7577-11e9-96cd-98fbf0ea17b3.png)

Also, as these CI questions come up often in Slack, let me know if you want to have an invite.

FeepingCreature commented:
> When this is merge, please make sure the commits are squashed.

Noted and aware. That's why it has that marker on the title.

> @FeepingCreature you are already a member of the Buildkite Organization.

Augh I keep forgetting. _slaps head_

> Why is void delegate() not callable?

_A delegate_ is callable. _`void delegate()`_, on the other hand, is not callable. `alias A = void delegate(); static assert(is(A()));`

wilzbach commented:
> why it has that marker on the title.

yYou might want to give draft PRs a try next time: https://github.blog/2019-02-14-introducing-draft-pull-requests/

JohanEngelen commented:
@FeepingCreature I very much appreciate the thoroughness. It is indeed a pity that there is currently no existing trait that would restore prior `alias` behaviour.
I don't understand how you would call `"a > b"` by just putting `()` after it?

FeepingCreature commented:
Right, you don't; it's more correct to say that strings like "a > b" happen to be the sort of thing that "would be passed to `std.functional.unaryFun`/`std.functional.binaryFun` so maybe the correct approach is checking for those explicitly.

jacob-carlborg commented:
> Noted and aware. That's why it has that marker on the title.

It would be more clear if you created fixup commits instead. There are quite a few contributors that don't know that commits should be amended or squashed. I can't keep track of which contributors know this and which don't, therefore I'm pointing it out as I see it.

FeepingCreature commented:
I'm only holding back on that because I don't know whether we want the flag or not yet, so I don't want to destroy the commit that had the flag until it's time to push for merging. But thanks for the info.

wilzbach commented:
If it passes on all CIs including the project tester, I don't think we need the special flag and can make Ethan immediately happy with the next release.

GooberMan commented:
Yee. And just chatting on IRC about it, it might be best to do a blog post or something with the release and explain that your code that fails to compile can be fixed by deleting lines/inserting template constraints/etc.

wilzbach commented:
> it might be best to do a blog post or something with the release

FYI @FeepingCreature et.al. Mike Parker compiles the release blog post based on changelog entries he sees. So just make sure that the changelog entry is long and detail enough, s.t. it does the following:

> explain that your code that fails to compile can be fixed by deleting lines/inserting template constraints/etc.

Though as far as I can tell the changelog entry looks pretty complete and detailed to me. @GooberMan anything that you miss in there as you brought this up?

FeepingCreature commented:
Flag removed. Ready on my end.

GooberMan commented:
@wilzbach - Chanelog looks good to me. Although I still expect a bit of initial confusion.

thewilsonator commented on test/compilable/ddocunittest.d:
```diff
@@ -1,4 +1,4 @@
-// PERMUTE_ARGS: -unittest
+// PERMUTE_ARGS: -unittest -preview=allowTypesForTemplateAlias
```
@FeepingCreature Still need to remove this use of flag.

FeepingCreature replied:
fixed

thewilsonator commented on test/compilable/test9029.d:
```diff
@@ -0,0 +1,19 @@
+// REQUIRED_ARGS: -preview=allowTypesForTemplateAlias
```
And this one.

FeepingCreature replied:
fixed thanks

thewilsonator reviewed (approved):
Otherwise LGTM.

wilzbach commented on test/compilable/test9029.d:
```diff
@@ -0,0 +1,17 @@
+enum NameOf(alias S) = S.stringof;
```
All test files should have a `TEST_OUTPUT` section (even if empty) to avoid that deprecations or warnings are silently ignored.

FeepingCreature replied:
fixed

GooberMan commented:
https://www.youtube.com/watch?v=z_5J1csi1AU

(The Dance of Joy for those that don't follow links)

TurkeyMan commented:
Wow... that just happened.
Who wants to do a PR to remove read/write access from `shared`? ;)

FeepingCreature commented:
I wish somebody would make a PR to remove `shared` and salt the earth...

TurkeyMan commented:
O_o

quickfur commented:
This PR has caused a regression: https://issues.dlang.org/show_bug.cgi?id=19884
